### ISSUE TITLE
HEAD-request hangs when served with body

### ISSUE THREAD

**Issue by Pyppe** [http4s/http4s#395]
Handling a HEAD-request with response-body hangs the connection. An example:
```scala
val service = HttpService {
  // Works as expected
  case HEAD -> Root / "works" -> Ok()
  // Keeps connection open
  case HEAD -> Root / "hangs" -> Ok("with content")
}
```
I would argue that ideally both handlers should yield the same response.

**Comment by rossabaker** [comment]
Which server backend?

**Comment by Pyppe** [comment]
`Http4sServlet` with `BlockingServletIo` (ver. 0.9.0)

**Comment by rossabaker** [comment]
Are you sending the HEAD request with `curl -X HEAD`?  The first command does as I expect.  The second command hangs.

```
% curl --head http://localhost:8080/http4s/science/hangs
HTTP/1.1 200 OK
Date: Tue, 25 Aug 2015 03:34:44 GMT
Content-Type: text/plain;charset=utf-8
Content-Length: 12
Server: Jetty(9.3.2.v20150730)

% curl -X HEAD http://localhost:8080/http4s/science/hangs
^C
```

Blaze is a little different: it sends the body if the client reads it despite sending a HEAD request.

```
% curl --head http://localhost:8080/http4s/science/hangs
HTTP/1.1 200 OK
Content-Type: text/plain; charset=UTF-8
Content-Length: 12
Date: Tue, 25 Aug 2015 03:40:18 GMT

% curl -X HEAD http://localhost:8080/http4s/science/hangs
with content
```

I think this is a [curl bug](http://sourceforge.net/p/curl/bugs/694/), whose resolution I disagree with.  On both backends, a properly behaved HTTP client gets the correct result.

**Comment by Pyppe** [comment]
Yup, noticed the issue with curl. Tested with jQuery `$.ajax({type: 'HEAD', ...})` and it works fine.

Of course, one could argue that it would be better if the server closed the connection by itself before the body.

**Comment by rossabaker** [comment]
We don't get control over the connection in Jetty or Tomcat.  

We do have this control in Blaze.  @bryce-anderson, is there a reason we shouldn't?  I can't think why not, but the fact that something as mature as Jetty doesn't makes me fear that I'm missing something.

**Comment by bryce-anderson** [comment]
Its simple enough to do in blaze though I seem to take the stance of 'if you sent a body in a response to a HEAD request, your wish is my command'. As an alternative, this could be addressed with a middleware very easily for all backends.

**Comment by hvesalai** [comment]
I would like to see the middleware (or blaze) work as follows:

Any body to a response to a HEAD request is discarded (so that HEAD requests can be served using the same service that does e.g. GET). The Writer (and hence headers) are chosen normally, i.e. if the corresponding GET response would have had Content-Length or Transfer-Encoding, so will the HEAD response.

Likewise (and I'll make an other issue about this): if a service gives a Content-Length, but then doesn't write enought bytes, the connection should be closed and no subsequent request should be read from it.

**Comment by hvesalai** [comment]
I.e. This should work so that GET gets the body and HEAD silently discards it.

```
val service = HttpService {
  case GET -> Root / "hello" -> go // serves headers, entity-headers and entity-body
  case HEAD -> Root / "hello" -> go // serves headers, entity-headers, but no entity-body
}

def go = Ok("World")
```

**Comment by rossabaker** [comment]
Fixed by #487
